src/qutip_qip/vqa.py

```diff
@@ -0,0 +1,842 @@
+            return (-1j * self.operator.get_hamiltonian(angles)).expm()
+        else:
+            if len(angles) != 1:
+                raise ValueError(
+                    "Expected one angle for singly-parameterized"
+                    " Hamiltonian."
+                )
+            else:
+                return (-1j * angles[0] * self.operator).expm()
+
+    def get_unitary_derivative(self, angles):
+        """
+        Compute the derivative of the block's unitary with respect to its
+        free parameter, assuming it is of the form e^{-i * theta * H}
+        for a free parameter theta.
+
+        Parameters
+        ----------
+        angle: list of float
+            free parameters to take derivatives with respect to
+
+        Returns
+        -------
+        derivative: float
+        """
+        if self.is_unitary or self.is_native_gate:
+            raise ValueError(
+                "Can only take derivative of block specified "
+                "by Hamiltonians or ParameterizedHamiltonian instances."
+            )
+        if isinstance(self.operator, ParameterizedHamiltonian):
+            return self.get_unitary_frechet_derivative(angles)
+        else:
+            if len(angles) != 1:
+                raise ValueError(
+                    "Expected a single angle for non-"
+                    "ParameterizedHamiltonian instance."
+                )
+            return self.get_unitary(angles) * -1j * self.operator
+
+    def get_unitary_frechet_derivative(self, angles, term_index=0):
+        """
+        Compute the derivative of the block's unitary with respect
+        to a free parameter, using the Frechet derivative of
+        the exponential function.
+
+        Parameters
+        ----------
+        angles: list of float
+            the free parameters to take derivatives with respect to
+        term_index: int, optional
+            Index of Parameterized Hamiltonian term that specifies the matrix
+            direction in which to take the derivative.
+
+        Returns
+        -------
+        derivative: float
+        """
+        if self.is_unitary or self.is_native_gate:
+            raise ValueError(
+                "Can only take frechet derivative of block "
+                "specified by Hamiltonians"
+            )
+        if isinstance(self.operator, ParameterizedHamiltonian):
+            arg = -1j * self.operator.get_hamiltonian(angles)
+            direction = -1j * self.operator.p_terms[term_index]
+        else:
+            if len(angles) != 1:
+                raise ValueError(
+                    "Expected a single angle for non-"
+                    "ParameterizedHamiltonian instance."
+                )
+            arg = -1j * self.operator * angles[0]
+            direction = -1j * self.operator
+        return Qobj(
+            expm_frechet(arg, direction, compute_expm=False),
+            dims=direction.dims,
+        )
```

Comment:
This seems the only issue that remained. I think `get_unitary_frechet_derivative` can be removed and the single part that is used
```python
            arg = -1j * self.operator.get_hamiltonian(angles)
            direction = -1j * self.operator.p_terms[term_index]
```
can be moved directly to `get_unitary_derivative`?

Once this is cleaned. I believe we are ready to merge!

Reply:
Was just writing a comment here haha. I'll post below anyway

That's true. I was originally thinking maybe the user would want to define a custom function to generate the Jacobian during optimization, which could depend on this method. Technically the method would have output the same as `get_unitary_derivative` if the operator is a `Qobj` Hamiltonian, but it's probably sensible to require that this method should only be called for `VQABlock`s that use a `ParameterizedHamiltonian`, so we could merge the two methods.

Reply:
> I was originally thinking maybe the user would want to define a custom function to generate the Jacobian during optimization, which could depend on this method.

I see. Yes, custom derivatives could be useful. They can even overwrite the full `get_unitary_derivative` I guess? One can surely further explore that framework in future PRs.